File: packages/aap-android/src/main/kotlin/dev/aap/sdk/AAPAgent.kt
package dev.aap.sdk

import dev.aap.sdk.crypto.*
import kotlinx.coroutines.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * Main entry point for the AAP Android / JVM SDK.
 *
 * ```kotlin
 * val agent = AAPAgent(AAPAgentConfig(name = "alice.android"))
 *
 * // Register on the network
 * val identity = agent.register()
 *
 * // Real-time relay session — works across Android, iOS, CLI, web
 * val session = agent.connectViaRelay(
 *     handle     = "alice.android",
 *     mailboxUrl = "wss://mailbox.aap.dev"
 * )
 * session.send("REQUEST_DATA", mapOf("query" to JsonPrimitive("hello")))
 * ```
 */
class AAPAgent(private val config: AAPAgentConfig) {

    private var identity:     AgentIdentity? = null
    private var privateKeyHex: String?       = null

    private val http = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(30,    TimeUnit.SECONDS)
        .build()

    private val json = Json { encodeDefaults = true; ignoreUnknownKeys = true }
    private val JSON_MEDIA = "application/json".toMediaType()

    // ── Register ───────────────────────────────────────────────────────────────

    /** Generate keys and register this agent with the AAP Registry. */
    suspend fun register(): AgentIdentity = withContext(Dispatchers.IO) {
        val (priv, pub) = generateEd25519KeyPair()
        privateKeyHex   = priv

        val aapAddress = "aap://${config.name}"
        val timestamp  = System.currentTimeMillis()
        val nonce      = generateNonce()

        val bodyObj = buildJsonObject {
            put("aap_address",    aapAddress)
            put("public_key_hex", pub)
            put("endpoint_url",   "https://agent.${config.name.replace(".", "-")}.local")
            put("capabilities",   buildJsonArray { config.capabilities.forEach { add(it) } })
            put("owner_type",     "human")
            put("timestamp",      timestamp)
            put("nonce",          nonce)
        }

        val bodyData = bodyObj.toString().toByteArray()
        val sig      = ed25519Sign(bodyData, priv)

        val withSig  = buildJsonObject {
            bodyObj.forEach { k, v -> put(k, v) }
            put("signature", sig)
        }

        val request = Request.Builder()
            .url("${config.registryUrl}/v1/register")
            .post(withSig.toString().toRequestBody(JSON_MEDIA))
            .build()

        val response = http.newCall(request).execute()
        val body     = response.body?.string() ?: ""

        if (!response.isSuccessful) {
            throw AAPError.RegistrationFailed(body)
        }

        val parsed   = Json.parseToJsonElement(body).jsonObject
        val id       = AgentIdentity(
            did          = parsed["did"]?.jsonPrimitive?.content ?: "did:aap:${config.name}",
            aapAddress   = parsed["aap_address"]?.jsonPrimitive?.content ?: aapAddress,
            publicKeyHex = pub,
            name         = config.name
        )
        identity = id
        println("✅ [AAPAndroid] Registered: $aapAddress  DID: ${id.did}")
        id
    }

    // ── Connect via relay ──────────────────────────────────────────────────────

    /**
     * Connect to another agent (or wait for a peer) via the real-time relay.
     *
     * Both sides call [connectViaRelay] with the same [handle].
     * Works across Android, iOS, web browser, Node.js CLI — any platform
     * that supports the AAP relay protocol.
     *
     * @param handle     aap:// address or bare handle to rendezvous on
     * @param mailboxUrl WebSocket URL of the AAP mailbox server
     */
    suspend fun connectViaRelay(handle: String, mailboxUrl: String): AAPSession {
        val id   = identity    ?: throw AAPError.NotRegistered()
        val priv = privateKeyHex ?: throw AAPError.NotRegistered()

        val relay = AAPRelay(handle = handle, mailboxUrl = mailboxUrl)
        relay.connect(did = id.did)

        val session = AAPSession(
            localIdentity = id,
            privateKeyHex = priv,
            remoteAddress = handle,
            relay         = relay
        )
        session.handshakeViaRelay()
        return session
    }

    // ── Lookup ─────────────────────────────────────────────────────────────────

    /** Look up any agent by aap:// address. */
    suspend fun lookup(address: String): JsonObject = withContext(Dispatchers.IO) {
        val encoded  = java.net.URLEncoder.encode(address, "UTF-8")
        val request  = Request.Builder()
            .url("${config.registryUrl}/v1/lookup/$encoded")
            .build()
        val response = http.newCall(request).execute()
        val body     = response.body?.string() ?: "{}"
        Json.parseToJsonElement(body).jsonObject
    }

    val currentIdentity: AgentIdentity? get() = identity
}

File: packages/aap-swift/Sources/AAPKit/AAPAgent.swift
import Foundation
import Crypto

/// Main entry point for the AAP Swift SDK.
///
/// ```swift
/// import AAPKit
///
/// let agent = AAPAgent(config: AAPAgentConfig(name: "alice.ios"))
/// let identity = try await agent.register()
///
/// // Real-time relay session (iOS ↔ CLI ↔ Android ↔ web)
/// let session = try await agent.connectViaRelay(
///     handle: "alice.ios",
///     mailboxURL: URL(string: "wss://mailbox.aap.dev")!
/// )
/// try await session.send(actionType: "REQUEST_DATA", payload: ["query": .string("hello")])
/// ```
@MainActor
public final class AAPAgent: @unchecked Sendable {

    // MARK: - State

    private let config:      AAPAgentConfig
    private var identity:    AgentIdentity?
    private var privateKey:  String?   // hex-encoded Ed25519 private key

    // MARK: - Init

    public init(config: AAPAgentConfig) {
        self.config = config
    }

    /// Convenience initialiser — just pass the agent name.
    public convenience init(name: String, capabilities: [String] = []) {
        self.init(config: AAPAgentConfig(name: name, capabilities: capabilities))
    }

    // MARK: - Register

    /// Generate keys and register this agent with the AAP Registry.
    @discardableResult
    public func register() async throws -> AgentIdentity {
        let (privHex, pubHex) = generateEd25519KeyPair()
        self.privateKey = privHex

        let aapAddress = "aap://\(config.name)"
        let timestamp  = Int64(Date().timeIntervalSince1970 * 1000)
        let nonce      = generateNonce()

        // Build body for signing (keys must match JS SDK ordering)
        let bodyDict: [String: Any] = [
            "aap_address":    aapAddress,
            "public_key_hex": pubHex,
            "endpoint_url":   "https://agent.\(config.name.replacingOccurrences(of: ".", with: "-")).local",
            "capabilities":   config.capabilities,
            "owner_type":     "human",
            "timestamp":      timestamp,
            "nonce":          nonce,
        ]

        // Sort keys so signature matches what the registry expects
        let sortedBody  = bodyDict.sorted { $0.key < $1.key }
        let bodyForSign = Dictionary(uniqueKeysWithValues: sortedBody)
        let bodyData    = try JSONSerialization.data(
            withJSONObject: bodyForSign,
            options: [.sortedKeys]
        )
        let signature   = try ed25519Sign(bodyData, privateKeyHex: privHex)

        // POST to registry
        var bodyWithSig = bodyForSign
        bodyWithSig["signature"] = signature

        var request        = URLRequest(url: config.registryURL.appendingPathComponent("v1/register"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody   = try JSONSerialization.data(withJSONObject: bodyWithSig, options: .sortedKeys)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 201 else {
            let msg = String(data: data, encoding: .utf8) ?? "unknown"
            throw AAPError.registrationFailed(msg)
        }

        let json = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        let id   = AgentIdentity(
            did:          json["did"]         as? String ?? "did:aap:\(config.name)",
            aapAddress:   json["aap_address"] as? String ?? aapAddress,
            publicKeyHex: pubHex,
            name:         config.name
        )
        self.identity = id
        print("✅ [AAPKit] Registered: \(aapAddress)  DID: \(id.did)")
        return id
    }

    // MARK: - Connect via relay

    /// Connect to another agent (or wait for a peer) via the real-time relay.
    ///
    /// Both sides call `connectViaRelay` with the same `handle`.
    /// The first caller becomes the host (SESSION_WAITING), the second becomes
    /// the guest (SESSION_READY broadcasts to both). Works across iOS, Android,
    /// web, CLI — any platform that supports the relay protocol.
    public func connectViaRelay(
        handle:     String,
        mailboxURL: URL
    ) async throws -> AAPSession {
        guard let identity, let privateKey else { throw AAPError.notRegistered }

        let relay = AAPRelay(handle: handle, mailboxURL: mailboxURL)
        try await relay.connect(did: identity.did)

        let session = AAPSession(
            localIdentity: identity,
            privateKeyHex: privateKey,
            remoteAddress: handle,
            registryURL:   config.registryURL,
            relay:         relay
        )
        try await session.handshakeViaRelay()
        return session
    }

    // MARK: - Lookup

    /// Look up any agent by aap:// address.
    public static func lookup(
        address:     String,
        registryURL: URL = URL(string: "https://registry.aap.dev")!
    ) async throws -> [String: Any] {
        let encoded  = address.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? address
        let url      = registryURL.appendingPathComponent("v1/lookup/\(encoded)")
        let (data, _) = try await URLSession.shared.data(from: url)
        return (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
    }

    public var currentIdentity: AgentIdentity? { identity }
}
